### PR TITLE
fix: builtin runner insert issue

### LINKED ISSUE #262: Telescope builtin: prompt in normal mode instead of insert mode
after running `:Telescope builtin` and selecting some picker, prompt is in normal mode instead of being in insert mode

### PR DESCRIPTION
**Pull request by tami5**
Fixes #262

I don't like the fact that I added an additional parameter,

@doubleloop please tested by checking out this branch after doing git fetch

### CODE REVIEW

**Comment by Conni2461** [comment]
> I don't like the fact that I add an additional parameter,

Does `vim.fn.mode()` help? Right now insert should be nil because you haven't added the second parameter in line 175.

**Comment by tami5** [comment]
h

> > I don't like the fact that I add an additional parameter,
> 
> Does `vim.fn.mode()` help? Right now insert should be nil because you haven't added the second parameter in line 175.

Good catch, hmm let me see, this function return a string that indicates the current mode. the close action actually kills the process along with any additional cmd in `run_builtin` so I'm not sure.

**Comment by tami5** [comment]
This case is tricky since the run_builtin doesn't renew the content of the float but rather close it then open a fresh new one. action.close insure that the when the window close it go to normal mode.

@Conni2461 I don't know do you have better idea to skip the additional param, or should we keep ?? since it doesn't bother any other picker :laughing:

**Comment by Conni2461** [comment]
yeah i think `mode()` will not work. Adding a new parameter in lua isn't that bad because it will be nil if it is not present. Let's wait for tj review and what he thinks

**Comment by tjdevries** [comment]
I think for the `builtin`, what we can do is give `action.close` a `post` action of `startinsert` again, which will put you back in insert mode.

Does that make sense? @Conni2461 knows about the `post` stuff, and can link to wherever we just used it :)

**Comment by tami5** [comment]
@Conni2461 how I can I use post stuff tj mentioned ? or should I just merge since its just a minor thing? Thanks

**Comment by Conni2461** [comment]
Don't merge. We will fix this. I am kinda here and can walk you through.
You wanna look at `enhance`. You can find docs in the readme and examples in `builtin.tags` and `builtin.current_buffer_fuzzy_find`.

**Comment by tami5** [comment]
> Don't merge. We will fix this. I am kinda here and can walk you through.
> You wanna look at `enhance`. You can find docs in the readme and examples in `builtin.tags` and `builtin.current_buffer_fuzzy_find`.
@Conni2461 
Got it

So I couldn't understand how I can extend with replace?

```
    attach_mappings = function(_)
      actions.goto_file_selection_edit:replace(actions.run_builtin)
      return true
    end

      actions.run_builtin:enhance {
        post = function()
          vim.cmd [[startinsert]]
        end
      }


```

**Comment by Conni2461** [comment]
I think tj suggested you should extend `actions.close`. And you would need to put this inside attach_mappings after the replace

**Comment by tami5** [comment]
@Conni2461 
Like that?
```lua
      actions.goto_file_selection_edit:replace(actions.run_builtin)
      actions.close:extend{
        post = function() vim.cmd [[startinsert]] end
      }
        return true
```

got `Error executing lua ...packer/opt/telescope.nvim/lua/telescope/builtin/init.lua:559: attempt to call method 'extend' (a nil value)`

**Comment by tami5** [comment]
thanks so much @Conni2461  please review

File: lua/telescope/actions/init.lua [hunk 1]
```diff
@@ -1,4 +1,4 @@
--- Actions functions that are useful for people creating their own mappings.
+jjjjj-- Actions functions that are useful for people creating their own mappings.
```
Comment (Conni2461) [inline lua/telescope/actions/init.lua:1]: :laughing:

File: lua/telescope/builtin/init.lua [hunk 2]
```diff
@@ -556,9 +556,12 @@ builtin.builtin = function(opts)
...
       actions.goto_file_selection_edit:replace(actions.run_builtin)
-      return true
-    end
-  }):find()
+      actions.close:enhance{
+        post = function() vim.cmd [[startinsert]] end
+      }
+        return true
+      end
+    }):find()
```
Comment (Conni2461) [inline lua/telescope/builtin/init.lua:564]: indent doesn't look right. Apart from that. LGTM

File: lua/telescope/actions/init.lua (hunk 1 above)
Comment (tami5) [reply lua/telescope/actions/init.lua:1]: @Conni2461 hhhhhhhhhhhhhhhhh forgot that i deleted stopinsert and that happens :D

File: lua/telescope/builtin/init.lua (hunk 2 above)
Comment (tami5) [reply lua/telescope/builtin/init.lua:564]: Nice thanks

**Comment by Conni2461** [comment]
I love that action api. Does this new version fix the bug?

**Comment by tami5** [comment]
Hmm not quite, it only works if I am in normal mode on I hit enter, while in insert mode and enter the picker goes back to normal mode @Conni2461 same issue

**Comment by Conni2461** [comment]
We could also replace `actions.close()` and do what we actually wanna do. Just an idea. i haven't looked into that issue

**Comment by tami5** [comment]
@Conni2461 Yeb but endup with duplication. I think I will look at it again tom.

**Comment by tami5** [comment]
Fixed with #287

**Comment by cempassi** [comment]
I found a way to go back in insert after closing/opening a builtin from lua with `vim.api.nvim_feedkeys('i', 'n', false)`. Maybe that can solve this problem for now?

**Comment by tami5** [comment]
@cempassi it did, but apparently there is one more issue left with builtin runner picker. backspace on clear prompt brakes the sorter

**Comment by tami5** [comment]
Wrong solution fixed by #300

**Comment by tjdevries** [comment]
Hmm, I don't like having the bonus parameter. All `actions` have had exactly the same signature up until this point. I can try and take a look at this later.

**Comment by tami5** [comment]
@tami5 yep me neither, couldn't find a solution that doesn't create undesired side effect. I'm interested to see how you endup solving it without introducing new param

**Comment by tjdevries** [comment]
:+1:

**Pull request merged**